taichi/ui/common/window_base.cpp

```diff
@@ -0,0 +1,161 @@
+#include "taichi/ui/common/window_base.h"
+
+TI_UI_NAMESPACE_BEGIN
+
+WindowBase ::WindowBase(AppConfig config) : config_(config) {
+  glfw_window_ = create_glfw_window_(config_.name, config_.width,
+                                     config_.height, config_.vsync);
+  glfwSetWindowUserPointer(glfw_window_, this);
+  set_callbacks();
+  last_record_time_ = glfwGetTime();
+}
+
+void WindowBase::set_callbacks() {
+  glfwSetKeyCallback(glfw_window_, key_callback);
+  glfwSetCursorPosCallback(glfw_window_, mouse_pos_callback);
+  glfwSetMouseButtonCallback(glfw_window_, mouse_button_callback);
+
+  input_handler_.user_key_callbacks.push_back([&](int key, int action) {
+    if (action == GLFW_PRESS) {
+      events_.push_back({EventType::Press, button_id_to_name(key)});
+    } else if (action == GLFW_RELEASE) {
+      events_.push_back({EventType::Release, button_id_to_name(key)});
+    }
+  });
+  input_handler_.user_mouse_button_callbacks.push_back(
+      [&](int key, int action) {
+        if (action == GLFW_PRESS) {
+          events_.push_back({EventType::Press, button_id_to_name(key)});
+        } else if (action == GLFW_RELEASE) {
+          events_.push_back({EventType::Release, button_id_to_name(key)});
+        }
+      });
+}
+
+CanvasBase *WindowBase::get_canvas() {
+  return nullptr;
+}
+
+void WindowBase::show() {
+  ++frames_since_last_record_;
+
+  double current_time = glfwGetTime();
+
+  if (current_time - last_record_time_ >= 1) {
+    double FPS =
+        (double)frames_since_last_record_ / (current_time - last_record_time_);
+    std::string glfw_window_text =
+        config_.name + "  " + std::to_string(FPS) + " FPS";
+
+    glfwSetWindowTitle(glfw_window_, glfw_window_text.c_str());
+    last_record_time_ = current_time;
+    frames_since_last_record_ = 0;
+  }
+
+  glfwPollEvents();
+}
+
+bool WindowBase::is_pressed(std::string button) {
+  int button_id = buttom_name_to_id(button);
+  return input_handler_.keys[button_id] > 0;
+}
+
+bool WindowBase::is_running() {
+  return !glfwWindowShouldClose(glfw_window_);
+}
+
+void WindowBase::set_is_running(bool value) {
+  glfwSetWindowShouldClose(glfw_window_, !value);
+}
+
+std::tuple<float, float> WindowBase::get_cursor_pos() {
```

Comment:
pair?

Reply:
Since we imported GLM why not glm::vec2

Reply:
std::pair has implicit type casts to pybind types that we can utilize when exporting to python